File: sommelier-analytics-mvp/backend/app/schemas/sale.py
"""
Sale schemas for API validation
"""
from pydantic import BaseModel, Field
from typing import Optional
from datetime import datetime, date
from uuid import UUID
from decimal import Decimal


class SaleBase(BaseModel):
    """Base sale schema"""
    wine_id: UUID
    sale_date: date
    quantity: int = Field(..., gt=0)
    unit_price: Decimal = Field(..., gt=0)
    unit_cost: Optional[Decimal] = Field(None, ge=0)
    
    server_name: Optional[str] = Field(None, max_length=100)
    table_number: Optional[str] = Field(None, max_length=20)
    notes: Optional[str] = Field(None, max_length=500)


class SaleCreate(SaleBase):
    """Schema for creating a new sale"""
    restaurant_id: UUID
    
    @property
    def total_amount(self) -> Decimal:
        return self.unit_price * self.quantity


class SaleBulkCreate(BaseModel):
    """Schema for bulk creating sales (CSV upload)"""
    restaurant_id: UUID
    sales: list[SaleBase]


class SaleResponse(SaleBase):
    """Schema for sale response"""
    id: UUID
    restaurant_id: UUID
    total_amount: Decimal
    profit: Optional[Decimal] = None
    profit_margin: Optional[float] = None
    created_at: datetime
    
    class Config:
        from_attributes = True


class SaleListResponse(BaseModel):
    """Schema for paginated sale list"""
    sales: list[SaleResponse]
    total: int
    page: int
    page_size: int
    total_pages: int

File: sommelier-analytics-mvp/frontend/src/types/sale.ts
import { z } from 'zod';

// Zod Schemas
export const saleSchema = z.object({
  id: z.string().uuid(),
  restaurant_id: z.string().uuid(),
  wine_id: z.string().uuid(),
  sale_date: z.string(), // Date string (YYYY-MM-DD)
  quantity: z.number().int().positive(),
  unit_price: z.number().positive(),
  total_amount: z.number().positive(),
  unit_cost: z.number().nonnegative().nullable(),
  server_name: z.string().max(100).nullable(),
  table_number: z.string().max(20).nullable(),
  notes: z.string().max(500).nullable(),
  created_at: z.string().datetime(),
  profit: z.number().nullable(),
  profit_margin: z.number().nullable(),
});

export const saleFormSchema = z.object({
  wine_id: z.string().uuid('Please select a wine'),
  sale_date: z.string().min(1, 'Sale date is required'),
  quantity: z.coerce.number().int().positive('Quantity must be at least 1'),
  unit_price: z.coerce.number().positive('Price must be greater than 0'),
  unit_cost: z.coerce.number().nonnegative().optional(),
  server_name: z.string().max(100).optional(),
  table_number: z.string().max(20).optional(),
  notes: z.string().max(500).optional(),
});

export const saleListResponseSchema = z.object({
  sales: z.array(saleSchema),
  total: z.number(),
  page: z.number(),
  page_size: z.number(),
  total_pages: z.number(),
});

// TypeScript Types
export type Sale = z.infer<typeof saleSchema>;
export type SaleFormData = z.infer<typeof saleFormSchema>;
export type SaleListResponse = z.infer<typeof saleListResponseSchema>;

export interface SaleListParams {
  restaurant_id: string;
  start_date?: string;
  end_date?: string;
  wine_id?: string;
  page?: number;
  page_size?: number;
}
